### ISSUE TITLE
Set Value with Android like iOS

### ISSUE THREAD

**Issue by chandrakanth1984** [appium/appium#4864]
Hi,

We proposed Appium for UI Automation and we are able to reduce the time consuming by entering text in iOS using setValue method. Is there any method in Android as well because appium is throwing error like this method is not implemented.

Thanks,
Chandrakanth

**Comment by jlipps** [comment]
no, for Android we're already using the fastest method available in UiAutomator

**Comment by bootstraponline** [comment]
Pretty sure adb shell input text is faster than UiAutomator on android. Maybe we should expose this if it isn't already?

**Comment by jlipps** [comment]
that's a good idea. reopening as a feature request

**Comment by chandrakanth1984** [comment]
I agree with @bootstraponline because sendkeys for android first try to clear the text in the textbox before sending the text to element. If there is hint in the text box sendkeys will take more time when compared to textbox without hint.

@jlipps 
Thanks for added this as feature. Can we expect this feature in next version of appium?

Thanks,
Chandrakanth

**Comment by jlipps** [comment]
@chandrakanth1984 not necessarily, no. we'll move it into a feature milestone once we've decided it will be in the next release.

**Comment by jlipps** [comment]
(merged #4449 into this)

**Comment by bootstraponline** [comment]
I think this should be exposed by implementing [setValueImmediate](https://github.com/appium/appium/blob/84e84584042d0242704c8a98e92b87a7ac74ceda/lib/server/routing.js#L140) for android (as mentioned in #4449).

**Comment by jlipps** [comment]
:+1: for setValueImmediate

**Comment by ShiranDadon** [comment]
Can't wait! Beers on me if you ever get to TLV!:)

**Comment by vikasg52** [comment]
Hi Team,
Can Anyone of you help me with an example of setValue(); method. I tried this and this is throwing error:
MobileElement EmailPlaceHolder= (MobileElement)driver.findElement(By.xpath("//UIATextField[@value='Email']"));
	//EmailPlaceHolder.tap(1, 2);
	((MobileElement) EmailPlaceHolder).setValue("1600");

Error :
The method setValue(String) is undefined for the type MobileElement.

**Comment by jasonkissinger** [comment]
We would also appreciate something to improve performance on setting text fields.  We're currently seeing it taking several seconds to set text in a single field, and multiply that across many field entries in a test and across 100s of tests, it is really slowing down our test runs.  If we could get that below a second then that would help us a lot.

As a simple example to show this timing locally, we added a test to AndroidUIAutomatorTest in the https://github.com/appium/java-client project:

```java
 public void textPerformance() {
    driver.scrollToExact("Views").click();
    driver.findElementByAndroidUIAutomator("text(\"Controls\")").click();
    driver.findElementByAndroidUIAutomator("text(\"1. Light Theme\")").click();

    long start = System.currentTimeMillis();
    driver.findElementByAndroidUIAutomator("resourceId(\"com.example.android.apis:id/edit\")").sendKeys("test 123");
    System.out.print("time: " + (System.currentTimeMillis() - start));
   }
```

**Comment by ashwiniadke** [comment]
I am using 1.4.10 and facing issue with sendKeys(Keys.ENTER) where before sending enter appium is clearing the text which is not correct, because of this my entered text is getting cleared . I think this should be a bug instead of feature enhancement .

**Comment by ashwiniadke** [comment]
Do we have any updates on this issue? This is a blocker for us :(

**Comment by imurchie** [comment]
This is now in Appium 1.5, through `setValueImmediate`.

**Comment by ArvindAbi** [comment]
@bootstraponline or @jlipps or @imurchie , Sorry if I am calling wide audience, I have a similar scenario as said by @ShiranDadon in the issue [#4449](https://github.com/appium/appium/issues/4449).  I ll type a recipient in the EditText Field, there will be list of suggestion and select recipient from the suggestion. Then I ll type next recipient and follow the same. In that case the already typed recipients should not get removed. But for me when I use SendKeys(). The first time it types the recipient-1 and selects recipient one from suggestion. Then while typing recipient-2 it is clearing the recipient-1 from text field and typing as "recipient-1recipient-2" which is not doing any good. Can you guys help me on this?

Appium Version: 1.5.3 (Ara)
Appium java client 4.0.0

**Comment by jeangb** [comment]
Hello, I'm using Appium  v1.4.16 on an android native application. My java client appium version is 4.1.2.
I can't see this **setValueImmediate** function, can you help me ?
The sendkeys lasts something like 5 seconds, it's way too long ...

**Comment by imurchie** [comment]
I think for Java it is just `setValue`.

**Comment by jeangb** [comment]
When I use setValue, appium server answers says that it's not implemented.

**Comment by imurchie** [comment]
Can you provide the server logs (as a [gist](https://gist.github.com) so that I can see what it is doing on the server?

**Comment by ssynix** [comment]
Take a look at https://github.com/appium/java-client/issues/427. I believe it's not available pre-1.5.

**Comment by jeangb** [comment]
@imurchie : thank you for your answer.
Here's the logs when I tried to use setValue : https://gist.github.com/jeangb/f779ae14ad311aeddc25f9817d699642

@ssynix : Thanks. Well I can't find Appium 1.5 for windows

**Comment by imurchie** [comment]
Yes, it appears it is not available pre-1.5. You can install 1.5. with `npm install -g appium@1.5`.

**Comment by jeangb** [comment]
@imurchie Thanks, setValue works with appium 1.5.
This is way faster than sendKeys :)
